File: 백준/Silver/2579. 계단 오르기/계단 오르기.cc
#include <iostream>
#define MAX 301
using namespace std;
int stair[MAX] = {0,};
int DP[MAX] = {0,};

int main(){
    int N;
    cin >> N;
    for (int i = 0; i < N; i++) cin >> stair[i];

    if (N <= 2){
        int sum = 0;
        for (int i = 0; i < N; i++){
            sum += stair[i];
        }
        cout << sum;
    }
    else{
        for (int i = 0; i < N; i++){
            DP[i] = max(DP[i - 2] + stair[i], DP[i - 3] + stair[i] + stair[i - 1]);
        }
        cout << DP[N - 1];
    }

    return 0;
}

File: 백준/Silver/2579. 계단 오르기/계단 오르기.py
N = int(input())
stair = [int(input()) for _ in range(N)]

sum = 0
if N <= 2:
    for i in range(N):
        sum += stair[i]
    print(sum)

else:
    dp = []
    dp.append(stair[0])
    dp.append(stair[0] + stair[1])
    dp.append(max(stair[2] + stair[0], stair[2] + stair[1]))

    for i in range(3, N):
        dp.append(max(dp[i-2] + stair[i], dp[i-3] + stair[i-1] + stair[i]))
    print(dp[N-1])